test/widgets/message_list_test.dart

```diff
@@ -888,6 +888,106 @@ void main() {
+
+          final opacity = tester.widget<AnimatedOpacity>(find.descendant(
+            of: find.byType(MarkAsReadWidget),
+            matching: find.byType(AnimatedOpacity)));
+          check(opacity.opacity).equals(expected ? 0.5 : 1.0);
+        }
+
+        void checkHasOpacityValue(WidgetTester tester, Widget child, double expectedOpacity) {
+          check(find.byWidget(child).evaluate()).length.equals(1);
+          check(tester.widget<AnimatedOpacity>(find.byType(AnimatedOpacity)).opacity).equals(expectedOpacity);
+        }
+
+        testWidgets('loading is changed correctly', (WidgetTester tester) async {
+          final narrow = TopicNarrow.ofMessage(message);
+          await setupMessageListPage(tester,
+            narrow: narrow, messages: [message], unreadMsgs: unreadMsgs);
+          check(isMarkAsReadButtonVisible(tester)).isTrue();
+
+          connection.prepare(
+            delay: const Duration(milliseconds: 2000),
+            json: UpdateMessageFlagsForNarrowResult(
+              processedCount: 11, updatedCount: 3,
+              firstProcessedId: null, lastProcessedId: null,
+              foundOldest: true, foundNewest: true).toJson());
+
+          checkAppearsLoading(tester, false);
+
+          await tester.tap(find.byType(MarkAsReadWidget));
+          await tester.pump();
+          checkAppearsLoading(tester, true);
+
+          await tester.pump(const Duration(milliseconds: 2000));
+          checkAppearsLoading(tester, false);
+        });
+
+        testWidgets('loading is changed correctly if request fails', (WidgetTester tester) async {
+          final narrow = TopicNarrow.ofMessage(message);
+          await setupMessageListPage(tester,
+            narrow: narrow, messages: [message], unreadMsgs: unreadMsgs);
+          check(isMarkAsReadButtonVisible(tester)).isTrue();
+
+          connection.prepare(httpStatus: 400, json: {
+            'code': 'BAD_REQUEST',
+            'msg': 'Invalid message(s)',
+            'result': 'error',
+          });
+
+          checkAppearsLoading(tester, false);
+
+          await tester.tap(find.byType(MarkAsReadWidget));
+          await tester.pump();
+          checkAppearsLoading(tester, true);
+
+          await tester.pump(const Duration(milliseconds: 2000));
+          checkAppearsLoading(tester, false);
+        });
+
+        testWidgets('in idle state', (WidgetTester tester) async {
+          final child = Container();
+
+          await tester.pumpWidget(MaterialApp(
+            home: MarkAsReadAnimation(hidden: false, loading: false, child: child)));
+
+          checkHasOpacityValue(tester, child, 1);
+        });
+
+        testWidgets('in loading state', (WidgetTester tester) async {
+          final child = Container();
+
+          await tester.pumpWidget(MaterialApp(
+            home: MarkAsReadAnimation(hidden: false, loading: true, child: child)));
+
+          checkHasOpacityValue(tester, child, 0.5);
+        });
```

Comment:
Following up on https://github.com/zulip/zulip-flutter/pull/710#discussion_r1692348196 : as written, this is a test that doesn't give us a lot of value, because it's operating at the layer of an implementation detail of how the UI is organized and as a result it largely just repeats what's in the code of MarkAsReadAnimation.

In the context of this section on testing philosophy that I linked earlier:
https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing
these are "unit tests" of the kind that doesn't serve us well.

Instead, the tests should be written in a way that's more in terms of the UI as the user experiences it. That means that instead of these four test cases that use MarkAsReadAnimation directly, the relevant behavior should be tested in the style of the two test cases earlier in this group: use `setupMessageListPage` to create a message-list page where the button appears, then tap on the button and check how the button looks.

(Some of these test cases may be redundant with what those other two test cases already do, in which case they can just be left out.)